scopesim/source/source.py

```diff
@@ -505,93 +538,77 @@ def shift(self, dx=0, dy=0, layers=None):
+        # new_source.spectra = deepcopy(self.spectra)
         for field in self.fields:
-            if isinstance(field, (fits.ImageHDU, fits.PrimaryHDU)) \
-                    and field._file is not None:  # and field._data_loaded is False:
-                new_source.fields.append(field)
-            else:
-                new_source.fields.append(deepcopy(field))
+            new_source.fields.append(deepcopy(field))
+            # TODO: The code below refers to DataContainer??
+            # if isinstance(field, (fits.ImageHDU, fits.PrimaryHDU)) \
+            #         and field._file is not None:  # and field._data_loaded is False:
+            #     new_source.fields.append(field)
+            # else:
+            #     new_source.fields.append(deepcopy(field))
 
         return new_source
 
     def append(self, source_to_add):
+        if not isinstance(source_to_add, Source):
+            raise ValueError(f"Cannot add {type(source_to_add)} object to Source object")
+
         new_source = source_to_add.make_copy()
         # If there is no field yet, then self._meta_dicts contains a
         # reference to self.meta, which is empty. This ensures that both are
         # updated at the same time. However, it is important that the fields
         # and _meta_dicts match when appending sources.
         if len(self.fields) == 0:
-            assert self._meta_dicts == [{}]
+            assert self._meta_dicts == [{}] or self._meta_dicts == []
             self._meta_dicts = []
-        if isinstance(source_to_add, Source):
-            for field in new_source.fields:
-                if isinstance(field, Table):
-                    field["ref"] += len(self.spectra)
-                    self.fields.append(field)
-
-                elif isinstance(field, (fits.ImageHDU, fits.PrimaryHDU)):
-                    if ("SPEC_REF" in field.header and
+
+        specrefoffset = max(self.spectra.keys()) + 1 if self.spectra else 0
+        for field in new_source.fields:
+            if isinstance(field, TableSourceField):
+                field.field["ref"] += specrefoffset
+                self.fields.append(field)
+
+            elif isinstance(field, HDUSourceField):
+                if ("SPEC_REF" in field.header and
                         isinstance(field.header["SPEC_REF"], int)):
-                        field.header["SPEC_REF"] += len(self.spectra)
-                    self.fields.append(field)
-                self.spectra += new_source.spectra
+                    field.field.header["SPEC_REF"] += specrefoffset
+                self.fields.append(field)
 
-                self._meta_dicts += source_to_add._meta_dicts
-        else:
-            raise ValueError(f"Cannot add {type(new_source)} object to Source object")
+            field.spectra = {k + specrefoffset: v for k, v in
+                             new_source.spectra.items()}
+            self._meta_dicts += source_to_add._meta_dicts
```

Comment:
So.. this code has been just wrong, because it is ran for every field. So if you have a source with two fields (so two _meta_dicts), and appended another source with two fields (also two _meta_dicts), you would have ended up with a source with 6 _meta_dicts, which doesn't make sense. So to fix that we should reduce the indentation.
```suggestion
        self._meta_dicts += source_to_add._meta_dicts
```
but since this code never worked, and is not allowed anymore in this new scheme, I propose to just remove it:
```suggestion
```
(Can github handle two suggestions in the same comment?)